### PR TITLE
[JENKINS-37557] Better support for pipelines

### PR DESCRIPTION
**Pull request by Typz**
A very strong limit to using from pipeline is that "contextual" environment variable (set during the job) cannot be accessed from the SCM [1]. To overcome this, this allow explicitely passing the environment variables from the checkout step.

It is now possible to use ssh agent or withEnv() block with repo checkout in Jenkins pipelines, like in the following code:

```groovy
sshAgent(['gerrit-ssh-credentials-id']) {
  checkout([$class: 'RepoScm', manifestRepositoryUrl: 'ssh://server/project',
            extraEnvVars: getContext(hudson.EnvVars)])
}
```

The pull request also includes a few minor changes:
* Fix crash when used from pipeline and changelog is empty.
* Include revision for new projects in changes view.
* In detailed view, hide missing fields: e.g. revision, author and commiter for new or removed projects.
* In compact view, show project name instead of 'unknown' author for new or removed projects.

[1]: https://issues.jenkins-ci.org/browse/JENKINS-37557

### CODE REVIEW

**Comment by Typz** [comment]
@rsandell : Hello! Could you take please a look at this pull request? It allows using repo plugin from pipeline script in combination with sshAgent(), which we need in our environment. Thanks!

File: src/main/resources/hudson/plugins/repo/RepoChangeLogSet/digest.jelly [hunk 1]
```diff
@@ -17,9 +17,16 @@
...
-					-- <a href="${rootURL}/${cs.author.url}/">${cs.author}</a> /
-					<a href="changes#detail${loop.index}">detail</a> <br/>
+				    <j:out value="${cs.msgAnnotated}"/> —
+				    <j:choose>
+                        <j:when test="${cs.authorName != null}">
+                            <a href="${rootURL}/${cs.author.url}/">${cs.author}</a>
+                        </j:when>
+                        <j:otherwise>
+                            ${cs.path}
+					    </j:otherwise>
```
Comment (rsandell) [inline src/main/resources/hudson/plugins/repo/RepoChangeLogSet/digest.jelly]: This spacing is not very beautiful and makes it quite hard to read for a mere human like myself :)

File: src/main/resources/hudson/plugins/repo/RepoChangeLogSet/index.jelly [hunk 2]
```diff
@@ -21,14 +21,20 @@
...
 						Project: ${cs.path}<br/>
-						Revision: ${cs.revision}<br/>
-						Author: <a href="${rootURL}/${cs.author.url}/">${cs.author}</a>
-							&amp;lt;${cs.authorEmail}&amp;gt; on ${cs.authorDate}<br/>
-						Committer: <a href="${rootURL}/${cs.committer.url}/">${cs.committer}</a>
-							&amp;lt;${cs.committerEmail}&amp;gt; on ${cs.committerDate}<br/>
+						<j:if test="${cs.revision != null}">
+						    Revision: ${cs.revision}<br/>
+						</j:if>
+			            <j:if test="${cs.authorName != null}">
```
Comment (rsandell) [inline src/main/resources/hudson/plugins/repo/RepoChangeLogSet/index.jelly]: Also quite weird usage of space here.

**Comment by rsandell** [comment]
Please in the future don't force push commits. It makes re reviewing the PR very hard as I can't see what changes there are since I last reviewed.

**Comment by Typz** [comment]
> Please in the future don't force push commits. It makes re reviewing the PR very hard as I can't see what changes there are since I last reviewed.

No problem, I will do this from now on.

Anyway, thanks for the merge! Any idea when you could make a new release of the plugin?

**Comment by rsandell** [comment]
It is released, the wiki has been turned to read only so I couldn't update the changelog there, so it's moved to GitHub releases from now on https://github.com/jenkinsci/repo-plugin/releases

**Comment by Typz** [comment]
yes, I have since seen the release! thank you very much !

**Pull request merged**